File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/restore_service.dart
// #docregion
import 'package:angular2/core.dart';

@Injectable()
class RestoreService<T> {
  T _originalItem;
  T _currentItem;

  setItem(T item) {
    _originalItem = item;
    _currentItem = clone(item);
  }

  T getItem() {
    return _currentItem;
  }

  T restoreItem() {
    _currentItem = _originalItem;
    return getItem();
  }

  T clone(T item) {
    // super poor clone implementation
    return item.clone();
  }
}
// #enddocregion

File: public/docs/_examples/hierarchical-dependency-injection/ts/app/restore.service.ts
// #docregion
export class RestoreService<T> {
  originalItem: T;
  currentItem: T;

  setItem (item: T) {
    this.originalItem = item;
    this.currentItem = this.clone(item);
  }

  getItem (): T {
    return this.currentItem;
  }

  restoreItem (): T {
    this.currentItem = this.originalItem;
    return this.getItem();
  }

  clone (item: T): T {
    // super poor clone implementation
    return JSON.parse(JSON.stringify(item));
  }
}
// #enddocregion
